one-ui/src/main/java/com/marklogic/hub/oneui/controllers/EnvironmentController.java

```diff
@@ -90,31 +96,109 @@ public JsonNode getProject() {
 
     @RequestMapping(value = "/install", method = RequestMethod.POST)
     @ResponseBody
-    public JsonNode install(@RequestBody ObjectNode payload) throws Exception {
-        String originalDirectory = environmentService.getProjectDirectory();
+    public void install(@RequestBody ObjectNode payload) throws Exception {
+        String directory = payload.get("directory").asText("");
+
         final Exception[] dataHubConfigurationException = {null};
+        HubDeployStatusListener listener = getInstallListener(dataHubConfigurationException, false);
+
+        install(directory, listener, dataHubConfigurationException);
+    }
+
+    @RequestMapping(value = "/project-download", produces = "application/zip")
+    public void downloadProject(HttpServletRequest request, HttpServletResponse response) {
+        HubProject project = hubConfig.getHubProject();
+        response.setContentType(MediaType.APPLICATION_OCTET_STREAM_VALUE);
+        response.addHeader("Content-Disposition", "attachment; filename=datahub-project.zip");
+        try (OutputStream out = response.getOutputStream()) {
+            project.exportProject(out);
+            response.flushBuffer();
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to download project;cause: " + e.getMessage());
+        }
+    }
+
+    @RequestMapping(value = "/reset", method = RequestMethod.POST)
+    @ResponseBody
+    public JsonNode reset() {
+        environmentService.reset();
+        ObjectNode obj = mapper.createObjectNode();
+        obj.put("reset", true);
+        return obj;
+    }
+
+    @RequestMapping(value = "/project-info", method = RequestMethod.GET)
+    @ResponseBody
+    public JsonNode getProjectInfo() {
+        return environmentService.getProjectInfo();
+    }
+
+    @RequestMapping(value = "/project-upload", method = RequestMethod.POST)
+    @ResponseBody
+    public void uploadProject(@RequestParam("zipfile") MultipartFile uploadedFile) throws Exception {
+        final Exception[] dataHubConfigurationException = {null};
+        HubDeployStatusListener listener = getInstallListener(dataHubConfigurationException, false);
+
+        HubProject project = hubConfig.getHubProject();
+
+        //backup first
+        backupExistingFSProject(project, listener);
+
+        // delete contents from the current project folder
+        cleanExistingFSProject(project, listener);
+
+        // extract the uploaded & zipped project file into the current project folder
+        extractZipProject(project, uploadedFile, listener);
```

Comment:
Why require the controller to make 3 separate calls here? We want controllers to be as thin and as simple as possible. The controller doesn't care about all of these details - it just wants to say - Hey, I've got this uploadedFile, please make this the new project. A ProjectManager class - see below - can handle all the details of backing up an existing project and then extracting the new one. I think this class just wants to call a method named "replaceProject".

Reply:
I can create one new method to combine 3 calls together.